refactor docs - navigable and easy-to-add examples

wwwtyro opened the pull request.

wwwtyro commented:
@vicapow Hope you like it! :tada: 

![screenshot from 2015-12-17 19-27-59](https://cloud.githubusercontent.com/assets/929443/11886865/5eba4ba2-a4f4-11e5-8479-eaed1b770336.png)

vicapow commented on index.js:
```diff
@@ -19,215 +19,49 @@
+var titles = Object.keys(contents);
 
-var Docs = React.createClass({
+var sidebar = document.getElementById('sidebar');
```
I think we should use React to create the sidebar if only to be consistent with the the rest of the code.

vicapow commented on index.js:
```diff
@@ -19,215 +19,49 @@
-  },
+  menuitem.addEventListener('click', (function(t, c) {
+    return function() {
+      setLocationHash(t, {replace: true});
```
I haven't used it myself but I've heard [react-router](https://github.com/rackt/react-router) is good for this kind of thing

wwwtyro commented:
Thanks for the review! I'll try to fix it up tomorrow evening.

If you're at all dissatisfied with the direction I've taken here, I'm happy to refactor it so that it's more in line with what you envisioned - I realize it's a significant departure from the original.

vicapow commented on index.css:
```diff
@@ -0,0 +1,96 @@
+  height: 50px;
+  line-height: 50px;
+  box-sizing: border-box;
+  font-family: 'Yanone Kaffeesatz', sans-serif;
```
what do you think about `Helvetica Neue`, instead?

wwwtyro replied:
Know where I can source a `Helvetica Neue` web font from? I've not found a free one.

vicapow commented:
It's looking good so far! I do have a some feedback. still going through.

vicapow commented:
I've been trying to keep build commits of `bundle.js` separate from regular changes.

vicapow commented on index.css:
```diff
@@ -0,0 +1,96 @@
+  line-height: 50px;
+  box-sizing: border-box;
+  font-family: 'Yanone Kaffeesatz', sans-serif;
+  color: #00E077;
```
can you try `#1FBAD6`?

vicapow commented on index.html:
```diff
@@ -1,7 +1,32 @@
 <html>
 <body>
+
+  <link href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.0.0/styles/rainbow.min.css" rel="stylesheet" type="text/css">
```
I kind of like `tomorrow` better than `rainbow`.

vicapow commented on content/third-party-overlay/third-party-overlays.md:
```diff
@@ -0,0 +1,6 @@
+You can also create your own overlays. Overlays will be passed the
+following props [transparently](https://
+github.com/uber/react-map-gl/blob/master/src/map.react.js#L589) 
+provided they're a direct children of a `<MapGL/>` map.
```
One thing I forgot to mention is that the docs also need to be updated to the new, 0.6 API so for example, it's no longer the case that overlays have to be direct children of the `<MapGL/>` element.

wwwtyro commented:
Excellent, thanks for all the feedback. :+1: I'll try to get through these tomorrow.

wwwtyro commented:
I addressed some issues and rebased to separate out the bundle rebuild into a new commit. I'm still working on upgrading react-map-gl to 0.6.1, but I've hit a snag rendering the overlays (they don't appear, no errors - see https://github.com/uber/react-map-gl/pull/41/files#diff-386b72445554b0b2557c2637ec3aea34R95 for my attempt). Unfortunately I've run out of time this evening, and am travelling this weekend, so won't be able to start again until Monday.

Have a great weekend!

vicapow commented:
@wwwtyro awesome progress.

you just need to [swap latitude and longitude](https://github.com/uber/react-map-gl/#swapping-latlng-for-lnglat). Sorry about that. We changed the API when Mapbox decided to swap them too so you can blame them, lol.

```diff
diff --git a/content/scatterplot/scatterplot.js b/content/scatterplot/scatterplot.js
index 839fed4..5efdecb 100644
--- a/content/scatterplot/scatterplot.js
+++ b/content/scatterplot/scatterplot.js
@@ -33,7 +33,6 @@ var stamenMapStyle = require('../../common/stamen-map-style');
 var CodeSnippet = require('../../common/code-snippet.react');
 var Markdown = require('../../common/markdown');
 
-
 module.exports = React.createClass({
 
   getInitialState: function getInitialState() {
@@ -53,7 +52,7 @@ module.exports = React.createClass({
         startDragLngLat: null
       },
       locations: Immutable.fromJS(d3.range(2000).map(function _map() {
-        return [37.788 + wiggle(0.01), -122.408 + wiggle(0.01)];
+        return [-122.408 + wiggle(0.01), 37.788 + wiggle(0.01), ];
       }))
     };
   },
diff --git a/package.json b/package.json
index beb745a..ab506a8 100644
--- a/package.json
+++ b/package.json
@@ -9,6 +9,7 @@
     "browserify": "^12.0.1",
     "budo": "^6.1.0",
     "d3": "^3.5.8",
+    "global": "^4.3.0",
     "highlight.js": "^8.8.0",
     "immutable": "^3.7.6",
     "js-stylesheet": "0.0.1",
```

wwwtyro commented:
Thanks! Yes, figured that out in https://github.com/wwwtyro/react-map-gl/commit/bd1e3a1465dc62f64d1968341c95a2ff9fbbdf47 -- should have left a comment saying so, my bad. Currently working on getting react-router working. :)

vicapow commented:
oh, nice! both our bad's 😜

wwwtyro commented:
hahaha

wwwtyro commented:
Alright, this one is ready for another review (no hurry though, enjoy your break!). The actual wording of the docs probably needs an expert eye to make sure I'm getting things right for the 0.6.x API.

vicapow commented:
Thanks, @wwwtyro! This is all looking good so far. I forgot to mention earlier but can you make sure `npm run lint` passes before I accept? checkout: [uber-standard](https://github.com/uber/standard) for more context.

wwwtyro commented:
No problem - done!

vicapow commented:
@wwwtyro still seeing a lot of output from `npm run lint` but it's all only indentation related. Maybe it's something about your editor?

````sh
react-map-gl ((c610a79...)) -> rm -rf node_modules && npm install && npm run lint
Error: Use Uber JavaScript Standard Style (https://github.com/uber/standard)
  /Users/victorpowell/Documents/projects/react-map-gl/common/code-snippet.react.js:39:6: Expected indentation of 6 characters. (indent)
  /Users/victorpowell/Documents/projects/react-map-gl/common/code-snippet.react.js:40:6: Expected indentation of 6 characters. (indent)
...
````

wwwtyro commented:
Weird, maybe I'm running the linter wrong or something, I'm error-free right now. The file you're running against giving that error looks like [this](https://github.com/wwwtyro/react-map-gl/blob/refactor-docs/common/code-snippet.react.js)?

vicapow commented:
@wwwtyro Yep. The file looks the same. uber-standard (and standard) both enforce 2 space indentation. Maybe you have a miscellaneous `.eslintrc` mucking things up? uber-standard --format can also fix the indentation issues automatically:

````sh
react-map-gl ((c610a79...)) -> uber-standard common/markdown.js
Error: Use Uber JavaScript Standard Style (https://github.com/uber/standard)
  /Users/victorpowell/Documents/projects/react-map-gl/common/markdown.js:34:6: Expected indentation of 6 characters. (indent)
react-map-gl ((c610a79...)) -> uber-standard --format common/markdown.js
react-map-gl ((c610a79...)) -> uber-standard common/markdown.js
````

wwwtyro commented:
I've tried a number of things, but it keeps insisting on 4-space indentation. Looking at [this in uber-standard](https://github.com/uber/standard/blob/master/rc/.eslintrc.json#L228) and [this eslint doc](http://eslint.org/docs/rules/indent), it seems like it really does want 4 spaces. I'm kicking off a VM right now to redo all the things and make sure I'm not mucking something up.

vicapow commented:
@wwwtyro I don't think `indent: [2, 4]` is a range. Looking at the source code, https://github.com/eslint/eslint/blob/master/lib/rules/indent.js#L57 it seems the first argument is exactly the number of spaces it wants. Maybe try updating `uber-standard` for the project to the latest version and try again? Should be version `5.1.0`.

```js
npm install uber-standard@latest --save
```

I get slightly different output when I do that:

````sh
react-map-gl ((c610a79...)) -> npm run lint

> react-map-gl-docs@0.0.0 lint /Users/victorpowell/Documents/projects/react-map-gl
> uber-standard

Error: Use Uber JavaScript Standard Style (https://github.com/uber/standard)
  /Users/victorpowell/Documents/projects/react-map-gl/common/code-snippet.react.js:30:5: Expected indentation of 2 space characters but found 4. (indent)
  /Users/victorpowell/Documents/projects/react-map-gl/common/code-snippet.react.js:32:5: Expected indentation of 2 space characters but found 4. (indent)
````

I also think `.eslintrc` files [are resolved recursively through every parent directory](http://eslint.org/docs/user-guide/configuring.html) so maybe that's what's happening here.

wwwtyro commented:
The eslint docs say:

> It takes an option as the second parameter which can be "tab" for tab-based indentation or a positive number for space indentations.

err... are those the right docs?

I looked for an errant `.eslintrc` file but didn't find one. My VM should be ready to test this soon.

wwwtyro commented:
I also get different errors with 5.1.0, but it's still looking for 4 spaces (it looks like it's just more particular about what lines it wants indented properly).

I tried it on a VM, and it still wants 4 spaces. Not sure what else to check. :\ Is it possible you have an errant `.eslintrc` file somewhere? Like, if you copy react-map-gl into /tmp and lint it from there, uber-standard still gives errors indicating it wants 2-space indentation?

Sorry, I'm sure this is the last thing you wanted to spend time on today! :O

wwwtyro commented:
It looks like it's looking for an .editorconfig file to define the indentation: 

https://github.com/uber/standard/blob/master/index.js#L72

which is 2 in the .editorconfig file included in uber-standard.

editorconfig-get-indent isn't hosted on github, but I printed out what it found when looking for .editorconfig, and it didn't find anything, in which case it defaults to 4. Is it possible you have an .editorconfig file lurking somewhere in your tree above react-map-gl? As the noob here, I'm loath to point the finger at your env, but I'm not sure I can otherwise interpret the fresh VM experiment I ran. 

Either way, it seems that the intent is probably to use 2 spaces for indentation, so I'll do that for now.

vicapow commented:
@wwwtyro That's it! Nice thinking firing up a VM. Here's what my `.editorconfig` file looks like for reference. Could you use this for now until we fix this issue with [uber-standard](https://github.com/uber/standard)? I filed a ticket with uber-standard at: https://github.com/uber/standard/issues/49 and will try to dig into this when I get some time.

```
# EditorConfig: http://editorconfig.org

root = true

[*]
indent_style = space
indent_size = 2
end_of_line = lf
charset = utf-8
trim_trailing_whitespace = true
insert_final_newline = true
```

wwwtyro commented:
Whew, I was worried I was starting to be more annoying than useful! :D 

Roger wilco re: .editorconfig. :+1:

Thanks for your patience & the merge!

vicapow commented:
I went and added a typical `.editorconfig` to the project, both the master and gh-pages branch. That should at least prevent this from popping up for this project in the future. 

https://github.com/uber/react-map-gl/commit/9fd6416f8dc810fb0e86472cf7abe414ae433120
https://github.com/uber/react-map-gl/commit/39962459240c537854af8d52231706d3047471c6
